x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/role/TransportPutRoleAction.java

```diff
@@ -6,75 +6,36 @@
  */
 package org.elasticsearch.xpack.security.action.role;
 
-import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.action.ActionListener;
-import org.elasticsearch.action.ActionRequestValidationException;
 import org.elasticsearch.action.support.ActionFilters;
 import org.elasticsearch.action.support.TransportAction;
 import org.elasticsearch.common.inject.Inject;
 import org.elasticsearch.tasks.Task;
 import org.elasticsearch.transport.TransportService;
-import org.elasticsearch.xcontent.NamedXContentRegistry;
 import org.elasticsearch.xpack.core.security.action.role.PutRoleAction;
 import org.elasticsearch.xpack.core.security.action.role.PutRoleRequest;
 import org.elasticsearch.xpack.core.security.action.role.PutRoleResponse;
-import org.elasticsearch.xpack.core.security.authz.support.DLSRoleQueryValidator;
-import org.elasticsearch.xpack.security.authz.ReservedRoleNameChecker;
 import org.elasticsearch.xpack.security.authz.store.NativeRolesStore;
 
-import static org.elasticsearch.action.ValidateActions.addValidationError;
-
 public class TransportPutRoleAction extends TransportAction<PutRoleRequest, PutRoleResponse> {
 
     private final NativeRolesStore rolesStore;
-    private final NamedXContentRegistry xContentRegistry;
-    private final ReservedRoleNameChecker reservedRoleNameChecker;
 
     @Inject
-    public TransportPutRoleAction(
-        ActionFilters actionFilters,
-        NativeRolesStore rolesStore,
-        TransportService transportService,
-        NamedXContentRegistry xContentRegistry,
-        ReservedRoleNameChecker reservedRoleNameChecker
-    ) {
+    public TransportPutRoleAction(ActionFilters actionFilters, NativeRolesStore rolesStore, TransportService transportService) {
         super(PutRoleAction.NAME, actionFilters, transportService.getTaskManager());
         this.rolesStore = rolesStore;
-        this.xContentRegistry = xContentRegistry;
-        this.reservedRoleNameChecker = reservedRoleNameChecker;
     }
 
     @Override
     protected void doExecute(Task task, final PutRoleRequest request, final ActionListener<PutRoleResponse> listener) {
-        final Exception validationException = validateRequest(request);
-        if (validationException != null) {
-            listener.onFailure(validationException);
-        } else {
-            rolesStore.putRole(request, request.roleDescriptor(), listener.safeMap(created -> {
-                if (created) {
-                    logger.info("added role [{}]", request.name());
-                } else {
-                    logger.info("updated role [{}]", request.name());
-                }
-                return new PutRoleResponse(created);
-            }));
-        }
-    }
-
-    private Exception validateRequest(final PutRoleRequest request) {
```

Comment:
This is now handled in the roles store instead, so the logic can be shared between bulk and single inserts. The benefit of keeping it in the roles store is that we can then merge the failed validation items with successful inserts.